### PR TITLE
Add scalar systems to linearFV version of SIMPLE

### PR DESCRIPTION
**Pull request by GiudGiud**
- [x] kernels
- [x] solve object attributes
- [x] solve the systems
- [x] tests
- [x] documentation

### CODE REVIEW

**Comment by GiudGiud** [comment]
Should be ready for review. I'm happy to remove the 2 commits adding a Physics if this is distracting us from the critical path of adding these linear capabilities

File: modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+This term is described by $\nabla \cdot \left(\rho\vec{u} c_p T \right)$ present in the scalar equation conservation for an incompressible/weakly-compressible formulation.
+
+For FV, the integral of the advection term of scalar $C_i$ over a cell can be expressed as:
+
+\begin{equation}
+\int\limits_{V_C} \nabla \cdot \left(\rho\vec{u} C_i \right) dV \approx \sum\limits_f (\rho \vec{u}\cdot \vec{n})_{RC} C_if |S_f| \,
+\end{equation}
+
+where $C_if$ is the face value of the scalar concentration. An interpolation scheme (e.g. upwind) can be used to compute the face value. This kernel adds the face contribution for each face $f$ to the right hand side and matrix.
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:13]:
```suggestion
where $C_{if}$ is the face value of the scalar concentration. An interpolation scheme (e.g. upwind) can be used to compute the face value. This kernel adds the face contribution for each face $f$ to the right hand side and matrix.
```

File: modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+The face mass flux $(\rho \vec{u}\cdot \vec{n})_{RC}$ is provided by the [RhieChowMassFlux.md] object which uses pressure
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:15]:
```suggestion
The face mass flux $(\vec{u}\cdot \vec{n})_{RC}$ is provided by the [RhieChowMassFlux.md] object which uses pressure
```

File: modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# LinearFVScalarAdvection
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:1]: The scalar advection schemes don't use the density. We will need some modifications in the RC object to be able to obtain the face velocities (volumetric flow rates).

File: modules/navier_stokes/doc/content/source/physics/WCNSLinearFVScalarTransportPhysics.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+- \mathbf{v} is the advecting velocity
+- $k_i$ the i-th scalar diffusivity
+- $Q_i$ is the i-th precursor source
+- $\lambda_i$ is a reaction coefficient. It should be negative for a loss term
+
+The kernels created are:
+
+- [LinearFVScalarAdvection.md] for the scalar advection term
+- [LinearFVDiffusion.md] for the scalar diffusion term
+- [LinearFVSource.md] for the reaction terms
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/physics/WCNSLinearFVScalarTransportPhysics.md:25]:
```suggestion
- [LinearFVReaction.md] for the reaction terms
```

File: modules/navier_stokes/include/linearfvkernels/LinearFVScalarAdvection.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+#pragma once
+
+#include "LinearFVFluxKernel.h"
+#include "RhieChowMassFlux.h"
+#include "LinearFVAdvectionDiffusionBC.h"
+
+/**
+ * An advection kernel that implements the advection term for the enthalpy in the
+ * energy equation.
```
Comment (grmnptr) [inline modules/navier_stokes/include/linearfvkernels/LinearFVScalarAdvection.h:18]:
```suggestion
 * An advection kernel that implements the advection term for the passive scalar transport equation.
```

File: modules/navier_stokes/include/linearfvkernels/LinearFVScalarAdvection.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  virtual Real computeBoundaryMatrixContribution(const LinearFVBoundaryCondition & bc) override;
+
+  virtual Real computeBoundaryRHSContribution(const LinearFVBoundaryCondition & bc) override;
+
+  virtual void setupFaceData(const FaceInfo * face_info) override;
+
+protected:
+  /// The Rhie-Chow user object that provides us with the face velocity
+  const RhieChowMassFlux & _mass_flux_provider;
```
Comment (grmnptr) [inline modules/navier_stokes/include/linearfvkernels/LinearFVScalarAdvection.h:42]: Maybe we can just divide the mass flux by the face rho. Either in the RC object or in the kernel.

File: modules/navier_stokes/include/physics/WCNSLinearFVScalarTransportPhysics.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "WCNSFVScalarTransportPhysicsBase.h"
+
+/**
+ * Creates all the objects needed to solve the Navier Stokes scalar transport equations
+ * using the nonlinear finite volume weakly-compressible discretization (WCNSFV)
```
Comment (grmnptr) [inline modules/navier_stokes/include/physics/WCNSLinearFVScalarTransportPhysics.h:16]:
```suggestion
 * using the linear finite volume weakly-compressible discretization (WCNSFV)
```

File: modules/navier_stokes/include/physics/WCNSLinearFVScalarTransportPhysics.h [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ */
+class WCNSLinearFVScalarTransportPhysics : public WCNSFVScalarTransportPhysicsBase
+{
+public:
+  static InputParameters validParams();
+
+  WCNSLinearFVScalarTransportPhysics(const InputParameters & parameters);
+
+private:
+  // TODO Rename to linear or solver?
```
Comment (grmnptr) [inline modules/navier_stokes/include/physics/WCNSLinearFVScalarTransportPhysics.h:26]: Solver sounds good!

File: modules/navier_stokes/src/executioners/SIMPLESolve.C [hunk 9]
```diff
@@ -34,6 +34,17 @@ SIMPLESolve::SIMPLESolve(Executioner & ex)
...
+  if (_has_passive_scalar_systems)
+  {
+    for (auto system_i : index_range(_passive_scalar_system_names))
+    {
+      _passive_scalar_system_numbers.push_back(
+          _problem.linearSysNum(_passive_scalar_system_names[system_i]));
+      _passive_scalar_systems.push_back(
+          &_problem.getLinearSystem(_passive_scalar_system_numbers[system_i]));
+    }
+  }
```
Comment (grmnptr) [inline modules/navier_stokes/src/executioners/SIMPLESolve.C:47]:
```suggestion
  if (_has_passive_scalar_systems)
    for (auto system_i : index_range(_passive_scalar_system_names))
    {
      _passive_scalar_system_numbers.push_back(
          _problem.linearSysNum(_passive_scalar_system_names[system_i]));
      _passive_scalar_systems.push_back(
          &_problem.getLinearSystem(_passive_scalar_system_numbers[system_i]));
    }
```

File: modules/navier_stokes/src/executioners/SIMPLESolve.C [hunk 10]
```diff
@@ -365,6 +376,48 @@ SIMPLESolve::solve()
     converged = NS::FV::converged(ns_residuals, ns_abs_tols);
   }
 
+  // If we have passive scalar equations, solve them here. We assume the material properties in the
+  // Navier-Stokes equations do not depend on passive scalar, as they are passive, therefore we
```
Comment (grmnptr) [inline modules/navier_stokes/src/executioners/SIMPLESolve.C:380]:
```suggestion
  // Navier-Stokes equations do not depend on passive scalsar, as they are passive, therefore we
```

File: modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+{
+  const auto * const adv_bc = static_cast<const LinearFVAdvectionDiffusionBC *>(&bc);
+  mooseAssert(adv_bc, "This should be a valid BC!");
+
+  const auto boundary_value_matrix_contrib = adv_bc->computeBoundaryValueMatrixContribution();
+
+  // We support internal boundaries too so we have to make sure the normal points always outward
+  const auto factor = (_current_face_type == FaceInfo::VarFaceNeighbors::ELEM) ? 1.0 : -1.0;
+
+  return boundary_value_matrix_contrib * factor * _face_mass_flux * _current_face_area;
```
Comment (grmnptr) [inline modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:74]: Maybe dividing with the face density here should be good.

File: modules/navier_stokes/src/physics/WCNSLinearFVScalarTransportPhysics.C [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+                        "If the nonorthogonal correction should be used when computing the normal "
+                        "gradient, notably in the diffusion term.");
+  params.addParam<bool>(
+      "scalar_boundary_two_term_expansion",
+      false,
+      "Whether to use the two term boundary expansion for boundary conditions that allow it");
+  return params;
+}
+
+WCNSLinearFVScalarTransportPhysics::WCNSLinearFVScalarTransportPhysics(
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVScalarTransportPhysics.C:34]: Very nice!

**Review by grmnptr — CHANGES_REQUESTED** [review]
Nice work, just a few comments.

File: modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+This kernel adds the contributions of the scalar advection term to the matrix and right hand side of the scalar equation system for the finite volume SIMPLE segregated solver [SIMPLE.md].
+
+This term is described by $\nabla \cdot \left(\rho\vec{u} c_p T \right)$ present in the scalar equation conservation for an incompressible/weakly-compressible formulation.
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:5]: well that is an important detail I missed. I ll have to rework this

File: modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md (hunk 3 above)
Comment (GiudGiud) [reply modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:1]: very true. this should be fixed now. That was a big oversight on my part, what was I thinking

**Comment by GiudGiud** [comment]
should be fixed

File: modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+For FV, the integral of the advection term of scalar $C_i$ over a cell can be expressed as:
+
+\begin{equation}
+\int\limits_{V_C} \nabla \cdot \left(\vec{u} C_i \right) dV \approx \sum\limits_f ( \vec{u}\cdot \vec{n})_{RC} C_if |S_f| \,
+\end{equation}
+
+where $C_{if}$ is the face value of the scalar concentration. An interpolation scheme (e.g. upwind) can be used to compute the face value. This kernel adds the face contribution for each face $f$ to the right hand side and matrix.
+
+The face mass flux $(\vec{u}\cdot \vec{n})_{RC}$ is provided by the [RhieChowMassFlux.md] object which uses pressure
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:15]:
```suggestion
The volumetric face flux $(\vec{u}\cdot \vec{n})_{RC}$ is provided by the [RhieChowMassFlux.md] object which uses pressure
```

File: modules/navier_stokes/doc/content/source/physics/WCNSLinearFVScalarTransportPhysics.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+!equation
+\dfrac{\partial \phi_i}{\partial t} + \nabla \cdot (\phi_i \mathbf{v}) - \nabla \cdot (k_i \nabla \phi_i) - Q_i - \lambda_i \phi_i = 0
+
+where:
+
+- $\phi_i$ is the i-th scalar quantity
+- \mathbf{v} is the advecting velocity
+- $k_i$ the i-th scalar diffusivity
+- $Q_i$ is the i-th precursor source
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/physics/WCNSLinearFVScalarTransportPhysics.md:18]:
```suggestion
- $Q_i$ is the i-th scalar source
```

File: modules/navier_stokes/include/userobjects/RhieChowMassFlux.h [hunk 16]
```diff
@@ -44,8 +44,10 @@ class RhieChowMassFlux : public GeneralUserObject,
   static InputParameters validParams();
   RhieChowMassFlux(const InputParameters & params);
 
-  /// Get the face velocity (used in advection terms)
+  /// Get the face velocity times density (used in advection terms)
   Real getMassFlux(const FaceInfo & fi) const;
+  /// Get the face velocity (used in advection terms)
```
Comment (grmnptr) [inline modules/navier_stokes/include/userobjects/RhieChowMassFlux.h:49]:
```suggestion
  /// Get the volumetric face flux (used in advection terms)
```

File: modules/navier_stokes/include/userobjects/RhieChowMassFlux.h [hunk 17]
```diff
@@ -44,8 +44,10 @@ class RhieChowMassFlux : public GeneralUserObject,
...
+  Real getFaceVelocity(const FaceInfo & fi) const;
```
Comment (grmnptr) [inline modules/navier_stokes/include/userobjects/RhieChowMassFlux.h:50]:
```suggestion
  Real getVolumetricFaceFlux(const FaceInfo & fi) const;
```

File: modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+      "The rhie-chow user-object which is used to determine the face velocity.");
+  params += Moose::FV::advectedInterpolationParameter();
+  return params;
+}
+
+LinearFVScalarAdvection::LinearFVScalarAdvection(const InputParameters & params)
+  : LinearFVFluxKernel(params),
+    _mass_flux_provider(getUserObject<RhieChowMassFlux>("rhie_chow_user_object")),
+    _advected_interp_coeffs(std::make_pair<Real, Real>(0, 0)),
+    _face_velocity(0.0)
```
Comment (grmnptr) [inline modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:34]:
```suggestion
    _volumetric_face_flux(0.0)
```

File: modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+{
+  Moose::FV::setInterpolationMethod(*this, _advected_interp_method, "advected_interp_method");
+}
+
+Real
+LinearFVScalarAdvection::computeElemMatrixContribution()
+{
+  return _advected_interp_coeffs.first * _face_velocity * _current_face_area;
```
Comment (grmnptr) [inline modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:42]:
```suggestion
  return _advected_interp_coeffs.first * _volumetric_face_flux * _current_face_area;
```

File: modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+Real
+LinearFVScalarAdvection::computeNeighborMatrixContribution()
+{
+  return _advected_interp_coeffs.second * _face_velocity * _current_face_area;
```
Comment (grmnptr) [inline modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:48]:
```suggestion
  return _advected_interp_coeffs.second * _volumetric_face_flux * _current_face_area;
```

File: modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+{
+  const auto * const adv_bc = static_cast<const LinearFVAdvectionDiffusionBC *>(&bc);
+  mooseAssert(adv_bc, "This should be a valid BC!");
+
+  const auto boundary_value_matrix_contrib = adv_bc->computeBoundaryValueMatrixContribution();
+
+  // We support internal boundaries too so we have to make sure the normal points always outward
+  const auto factor = (_current_face_type == FaceInfo::VarFaceNeighbors::ELEM) ? 1.0 : -1.0;
+
+  return boundary_value_matrix_contrib * factor * _face_velocity * _current_face_area;
```
Comment (grmnptr) [inline modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:74]:
```suggestion
  return boundary_value_matrix_contrib * factor * _volumetric_face_flux * _current_face_area;
```

File: modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+LinearFVScalarAdvection::computeBoundaryRHSContribution(const LinearFVBoundaryCondition & bc)
+{
+  const auto * const adv_bc = static_cast<const LinearFVAdvectionDiffusionBC *>(&bc);
+  mooseAssert(adv_bc, "This should be a valid BC!");
+
+  // We support internal boundaries too so we have to make sure the normal points always outward
+  const auto factor = (_current_face_type == FaceInfo::VarFaceNeighbors::ELEM ? 1.0 : -1.0);
+
+  const auto boundary_value_rhs_contrib = adv_bc->computeBoundaryValueRHSContribution();
+  return -boundary_value_rhs_contrib * factor * _face_velocity * _current_face_area;
```
Comment (grmnptr) [inline modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:87]:
```suggestion
  return -boundary_value_rhs_contrib * factor * _volumetric_face_flux * _current_face_area;
```

File: modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+void
+LinearFVScalarAdvection::setupFaceData(const FaceInfo * face_info)
+{
+  LinearFVFluxKernel::setupFaceData(face_info);
+
+  // Caching the velocity on the face which will be reused in the advection term's matrix and right
+  // hand side contributions
+  _face_velocity = _mass_flux_provider.getFaceVelocity(*face_info);
```
Comment (grmnptr) [inline modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:97]:
```suggestion
  _volumetric_face_flux = _mass_flux_provider.getVolumetricFaceFlux(*face_info);
```

File: modules/navier_stokes/src/userobjects/RhieChowMassFlux.C [hunk 24]
```diff
@@ -234,6 +234,18 @@ RhieChowMassFlux::getMassFlux(const FaceInfo & fi) const
   return _face_mass_flux.evaluate(&fi);
 }
 
+Real
+RhieChowMassFlux::getFaceVelocity(const FaceInfo & fi) const
```
Comment (grmnptr) [inline modules/navier_stokes/src/userobjects/RhieChowMassFlux.C:238]:
```suggestion
RhieChowMassFlux::getVolumetricFaceFlux(const FaceInfo & fi) const
```

File: modules/navier_stokes/src/userobjects/RhieChowMassFlux.C [hunk 25]
```diff
@@ -234,6 +234,18 @@ RhieChowMassFlux::getMassFlux(const FaceInfo & fi) const
...
+{
+  const Moose::FaceArg face_arg{&fi,
+                                /*limiter_type=*/Moose::FV::LimiterType::CentralDifference,
+                                /*elem_is_upwind=*/true,
+                                /*correct_skewness=*/false,
+                                &fi.elem()};
+  const Real face_rho = _rho(face_arg, Moose::currentState());
+  return _face_mass_flux.evaluate(&fi) / face_rho;
```
Comment (grmnptr) [inline modules/navier_stokes/src/userobjects/RhieChowMassFlux.C:246]:
```suggestion
  return _face_mass_flux[fi.id()] / face_rho;
```

**Review by grmnptr — CHANGES_REQUESTED** [review]
Nice, almost there!

**Review by grmnptr — APPROVED** [review]
Nice!

**Comment by GiudGiud** [comment]
Thanks for the review!
@tanoret fyi you may want to rebase to get the latest version of the solve object

**Comment by GiudGiud** [comment]
Forgot to press enter:
test failures are unrelated. A distributed p2 failure in OpenMPI in unrelated tests, bluecrab not building due to an update in lowerD elements for Griffin that is unrelated as well

**Pull request merged**